xarray/core/dataset.py

```diff
@@ -419,6 +419,18 @@ def _replace_vars_and_dims(self, variables, coord_names=None,
             obj = self._construct_direct(variables, coord_names, dims, attrs)
         return obj
 
+    def _replace_indexes(self, indexes):
+        variables = OrderedDict()
+        for k, v in iteritems(self._variables):
+            if k in indexes.keys():
+                idx = indexes[k]
+                variables[k] = Coordinate(idx.name, idx)
+            else:
+                variables[k] = v
+        obj = self._replace_vars_and_dims(variables)
+        dim_names = {dim: idx.name for dim, idx in iteritems(indexes)}
+        return obj.rename(dim_names)
```

Comment:
Can we make the rename only done if necessary? I think this can be kind of expensive.

Putting things together:
```python
variables = self._variables.copy()
for name, idx in indexes.items():
   variables[name] = Coordinate(name, idx)
obj = self._replace_vars_and_dims(variables)

# switch from dimension to level names, if necessary
dim_names = {}
for dim, idx in indexes.items():
    if idx.name != dim:
        dim_names[dim] = idx.name
if dim_names:
    obj = obj.rename(dim_name)
```